pulp_container/app/downloaders.py

```diff
@@ -95,14 +100,47 @@ async def _run(self, handle_401=True, extra_data=None):
                         return await self._run(handle_401=False, extra_data=extra_data)
                 else:
                     raise
+
             to_return = await self._handle_response(response)
+
             await response.release()
             self.response_headers = response.headers
 
         if self._close_session_on_finalize:
             self.session.close()
         return to_return
 
+    def _ensure_writer_has_open_file(self):
+        """
+        Create a temporary file on demand.
+
+        Create a temporary file when it's actually used, allowing plugin writers to instantiate
+        many downloaders in memory.
+
+        This method sets the path of NamedTemporaryFile dynamically based on whether it is running
+        from a task or not. Otherwise, permission errors might be raised when Pulp is trying to
+        download a file from api-app and write to a user space.
+        """
+        if not self._writer:
+            dir_path = settings.WORKING_DIRECTORY if Task.current() is None else "."
+            self._writer = tempfile.NamedTemporaryFile(dir=dir_path, delete=False)
+            self.path = self._writer.name
+            self._digests = {n: pulp_hashlib.new(n) for n in Artifact.DIGEST_FIELDS}
+            self._size = 0
+
+    def fetch(self, extra_data=None):
+        """
+        Run the download synchronously with additional data and return the `DownloadResult`.
+
+        Returns:
+            :class:`~pulpcore.plugin.download.DownloadResult`
+                or :class:`~aiohttp.client.ClientResponse`
+
+        Raises:
+            Exception: Any fatal exception emitted during downloading
+        """
+        return asyncio.get_event_loop().run_until_complete(self.run(extra_data=extra_data))
+
```

Comment:
`fetch()` was copied here for the sake of CI because its updated version was not yet released back then.

Concerning `_ensure_writer_has_open_file()` (original version here: https://github.com/pulp/pulpcore/blob/3cb7ef4369fd612d17e2be7863845f6cc74c141f/pulpcore/download/base.py#L112), Pulp is by default downloading content from api-app to an inaccessible folder. Visit the docstring for more details.

Reply:
What is wrong with always using dir_path="."? The task should have cd'd into it's working directory (btw a subdirectory of settings.WORKING_DIRECTORY).

Reply:
We download the content from api-app. The task is not involved in pull-through caching. This is what I see without the change, just using the pulpcore's facility:
```
  File "/src/pulpcore/pulpcore/download/http.py", line 258, in download_wrapper
    return await self._run(extra_data=extra_data)
  File "/src/pulp_container/pulp_container/app/downloaders.py", line 95, in _run
    return await self._run(handle_401=False, extra_data=extra_data)
  File "/src/pulp_container/pulp_container/app/downloaders.py", line 104, in _run
    to_return = await self._handle_response(response)
  File "/src/pulpcore/pulpcore/download/http.py", line 209, in _handle_response
    await self.finalize()
  File "/src/pulpcore/pulpcore/download/base.py", line 166, in finalize
    self._ensure_writer_has_open_file()
  File "/src/pulpcore/pulpcore/download/base.py", line 130, in _ensure_writer_has_open_file
    self._writer = tempfile.NamedTemporaryFile(dir=".", suffix=suffix, delete=False)
  File "/usr/lib64/python3.8/tempfile.py", line 540, in NamedTemporaryFile
    (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "/usr/lib64/python3.8/tempfile.py", line 250, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
PermissionError: [Errno 13] Permission denied: './tmpg9hy1dqd-latest'
```

Reply:
Ah, i misread the logic. So you are saying the api-worker needs to start writing stuff to disk? This opens an new pandora box... (and probably some repercussions with SELinux)
Is there any reasonable way to move this out to the content-app?

Reply:
The problem with dispatching tasks from content-app is gone after calling `set_guid(generate_guid())`. Therefore, we can revive the idea of streaming the content from content-app directly. With this, we will likely not store the content within a repository. This is the subject for discussion.

Attaching the suggestion about a task that will handle blobs downloading in the background:
```
That task will take the artifact, read it, draft on-demand blobs from it and creates and adds a manifest to the repository. (The Manifest having healthy relationships to its blobs.)
In the meantime the user will probably download a certain number of the blobs in that manifest.
Again, pulp-content will stream download them to the client if not yet in pulp, and (maybe) create ephemeral blobs from them.
Once the task ran, we now have on-demand blobs, and the serving goes through the old channels of pulp.
```

Reply:
You see me confused. The fact that we can dispatch that task means that we _can_ add the content from requests solely talking to the content_app.